File: Hash-Based Searching/Open Addressing Search/Linear Probing/LinearProbing.py
TABLE_SIZE = 11
EMPTY = -1


class LinearProbingHashTable:

    def __init__(self):
        self.table = [EMPTY] * TABLE_SIZE

    def hash_function(self, key):
        return key % TABLE_SIZE

    def insert(self, key):

        index = self.hash_function(key)

        for i in range(TABLE_SIZE):

            pos = (index + i) % TABLE_SIZE

            if self.table[pos] == EMPTY:
                self.table[pos] = key
                return True

        return False

    def search(self, key):

        index = self.hash_function(key)

        for i in range(TABLE_SIZE):

            pos = (index + i) % TABLE_SIZE

            if self.table[pos] == EMPTY:
                return False

            if self.table[pos] == key:
                return True

        return False

    def display(self):

        print("\nHash Table")

        for i in range(TABLE_SIZE):

            if self.table[i] == EMPTY:
                print(f"[{i}] = EMPTY")
            else:
                print(
                    f"[{i}] = {self.table[i]}"
                )


if __name__ == "__main__":

    ht = LinearProbingHashTable()

    ht.insert(22)
    ht.insert(33)
    ht.insert(44)
    ht.insert(55)

    ht.display()

    print(
        "\nSearch 44:",
        "Found"
        if ht.search(44)
        else "Not Found"
    )

File: Hash-Based Searching/Open Addressing Search/Linear Probing/LinearProbing.java
public class LinearProbing {

    private static final int TABLE_SIZE = 11;
    private static final int EMPTY = -1;

    private final int[] table;

    public LinearProbing() {
        table = new int[TABLE_SIZE];

        for (int i = 0; i < TABLE_SIZE; i++) {
            table[i] = EMPTY;
        }
    }

    private int hashFunction(int key) {
        return key % TABLE_SIZE;
    }

    public boolean insert(int key) {

        int index = hashFunction(key);

        for (int i = 0; i < TABLE_SIZE; i++) {

            int pos =
                (index + i) % TABLE_SIZE;

            if (table[pos] == EMPTY) {
                table[pos] = key;
                return true;
            }
        }

        return false;
    }

    public boolean search(int key) {

        int index = hashFunction(key);

        for (int i = 0; i < TABLE_SIZE; i++) {

            int pos =
                (index + i) % TABLE_SIZE;

            if (table[pos] == EMPTY)
                return false;

            if (table[pos] == key)
                return true;
        }

        return false;
    }

    public void display() {

        System.out.println(
            "\nHash Table:"
        );

        for (int i = 0; i < TABLE_SIZE; i++) {

            System.out.print(
                "[" + i + "] = "
            );

            if (table[i] == EMPTY)
                System.out.println("EMPTY");
            else
                System.out.println(table[i]);
        }
    }

    public static void main(String[] args) {

        LinearProbing ht =
            new LinearProbing();

        ht.insert(22);
        ht.insert(33);
        ht.insert(44);
        ht.insert(55);

        ht.display();

        System.out.println(
            "\nSearch 44: "
            + (ht.search(44)
                ? "Found"
                : "Not Found")
        );
    }
}
